[Inductor] Expand dtype aware codegen for libdevice and tl.math ops

blaine-rister opened the pull request:
Summary: Previously, only `torch.sqrt` was dtype aware. This PR updates most of the unary ops to have it as well. This is necessary to get correct code for libdevice and tl.math ops, as their inputs need to be upcasted to float32.

Test Plan: Added CI tests for all the new ops.

Differential Revision: D65517197




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

blaine-rister commented:
This is not yet ready for review. I still need to run the CI test.

Skylion007 commented on test/inductor/test_torchinductor.py:
```diff
@@ -12311,19 +12311,27 @@ def func(a, b, c, d):
                 self.assertEqual(fp32_cast_in_code, upcast_to_fp32)
 
         @requires_gpu()
+        @parametrize("op", [torch.rsqrt, torch.sqrt, torch.isnan, torch.floor, torch.ceil, torch.tan, torch.atan, torch.atanh, torch.sigmoid, torch.log2, torch.log10, torch.cosh, torch.sinh, torch.acosh, torch.asinh, torch.asin, torch.acos, torch.asinh, torch.erf, torch.lgamma, torch.sin, torch.cos, torch.exp, torch.expm1, torch.exp2, torch.abs])
```
Can we use the opinfo db and select all pointwise ops or something programatically? I am worried this list will not be updated in the future as we add more ops.

Skylion007 replied:
Just iterate through the class attributes directly?

blaine-rister replied:
Do you mean we should iterate through all the methods of `TritonOverrides`? That could work but this test wouldn't necessarily apply to all of them. Wouldn't we need some extra info to figure out which ones generate `libdevice` calls to a unary op? It might be possible to get this with `inspect`.

blaine-rister commented on test/inductor/test_torchinductor.py:
```diff
@@ -12311,19 +12311,27 @@ def func(a, b, c, d):
+        def test_dtype_aware_codegen(self, op: typing.Callable, load_upcast_to_fp32, input_dtype):
+            """
+            Test dtype aware codegen for some libdevice calls.
+            Operands should be upcast to float32, and the output should be downcast to float16.
```
nit: or bfloat16

eellison commented on torch/_inductor/codegen/triton.py:
```diff
@@ -832,42 +845,35 @@ def constant(cls, value, dtype):
     @staticmethod
     def exp(x):
-        return f"tl_math.exp({x})"
+        return _maybe_upcast_float32("tl_math.exp", [x])
```
nit - would this be cleaner as a decorator ? 
```
@maybe_upcast_float32
@staticmethod
def libdevice_exp(x):
    return f"libdevice.exp({x})"
 
```

blaine-rister replied:
Very good point. Now that you mention it, in most cases it would be equivalent to just have a decorator that upcasts inputs and downcasts the output. I can try this out.

We might be able to inspect the decorator to automatically generate a test list as well, as @Skylion007 suggested.

eellison replied:
@blaine-rister I added tests here - https://github.com/pytorch/pytorch/pull/140057. we take all the opinfos for pointwise ops we have inductor lowerings of. let's follow up a bit more on other planning in workplace chat.

blaine-rister replied:
I changed things to the decorator in 
6364feb. This seems to be working well, although a few of the test cases are failing. It seems like float16/bfloat16 dtypes might not be propagating through all those cases, meaning the codegen doesn't end up casting up. I'll follow up offline.

blaine-rister committed:
[Inductor] Expand dtype aware codegen for unary ops (#139839)

Summary:

Previously, only `torch.sqrt` was dtype aware. This PR updates most of the unary ops to have it as well. This is necessary to get correct code for libdevice and tl.math ops, as their inputs need to be upcasted to float32.

Test Plan: Added CI tests for all the new ops.

Differential Revision: D65517197

blaine-rister commented on test/inductor/test_torchinductor.py:
```diff
@@ -12326,19 +12326,27 @@ def func(a, b, c, d):
+                return op(a * b * c * d)
 
             inps = (torch.rand((32, 32), device=GPU_TYPE, dtype=input_dtype),) * 4
+            op_str = op.__name__.removeprefix("torch.")
```
TODO does it actually have this prefix?

blaine-rister commented:
Superseded by https://github.com/pytorch/pytorch/pull/140864
